tip-distributor/src/lib.rs

```diff
@@ -460,104 +470,212 @@ pub fn derive_tip_distribution_account_address(
-            let semaphore = semaphore.clone();
-            async move {
-                let permit = semaphore.clone().acquire_owned().await.unwrap();
-                let res = match rpc_client.send_transaction(tx).await {
-                    Ok(sig) => {
-                        info!("sent transaction: {sig:?}");
-                        drop(permit);
-                        sleep(Duration::from_secs(10)).await;
-
-                        let _permit = semaphore.acquire_owned().await.unwrap();
-                        match rpc_client.confirm_transaction(&sig).await {
-                            Ok(true) => Ok(()),
-                            Ok(false) => Err(Error::new_with_request(
-                                ErrorKind::Custom("transaction failed to confirm".to_string()),
-                                RpcRequest::SendTransaction,
-                            )),
-                            Err(e) => Err(e),
-                        }
-                    }
-                    Err(e) => Err(e),
-                };
-
-                let res = res
-                    .err()
-                    .map(|e| {
-                        if let ErrorKind::TransactionError(AlreadyProcessed) = e.kind {
-                            Ok(())
-                        } else {
-                            error!("error sending transaction {sig:?} error: {e:?}");
-                            Err(e)
-                        }
-                    })
-                    .unwrap_or(Ok(()));
-
-                (*sig, res)
-            }
-        });
-
-        errors = futures::future::join_all(futs)
-            .await
+        let send_res = futures::future::join_all(send_futs).await;
+        let new_errors = send_res
             .into_iter()
-            .filter(|(sig, result)| {
-                if result.is_err() {
-                    true
-                } else {
-                    let _ = signatures_to_transactions.remove(sig);
-                    false
+            .filter_map(|(hash, txn, result)| match result {
+                Err(e) => Some((txn.signatures[0], e)),
+                Ok(..) => {
+                    let _ = transactions_to_process.remove(&hash);
+                    None
                 }
             })
-            .map(|(sig, result)| {
-                let e = result.err().unwrap();
-                warn!("error sending transaction: [error={e}, signature={sig}]");
-                (sig, e)
-            })
             .collect::<HashMap<_, _>>();
+
+        errors.extend(new_errors);
     }
 
-    errors
+    (transactions_to_process.values().cloned().collect(), errors)
+}
+async fn signed_send(
+    signer: &Keypair,
+    rpc_client: &RpcClient,
+    blockhash: Hash,
+    mut txn: Transaction,
+) -> (Transaction, solana_rpc_client_api::client_error::Result<()>) {
+    txn.sign(&[signer], blockhash); // just in time signing
+    let res = match rpc_client.send_and_confirm_transaction(&txn).await {
```

Comment:
how long does this wait to confirm?

Reply:
should try every 0.5 secs: https://github.com/jito-labs/jito-solana/blob/f6b2cc843b675653ecb9061d326e4e41b5b0dda1/rpc-client/src/nonblocking/rpc_client.rs#L708

Reply:
it looks like max amount of time would be blockhash expiry?

Reply:
yea looks like that